app/controllers/provider/admin/cms/visit_portal_controller.rb

```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
+
+class Provider::Admin::CMS::VisitPortalController < Provider::Admin::CMS::BaseController
+  # Encrypt the CMS token under a temporary SSO token and redirect to the Developer Portal
+  def with_token
+    verifier = Rails.application.message_verifier(:cms_token)
+    cms_token = current_account.settings.cms_token!
+    expires_at = Time.now.utc.round + 1.minute
+    signing_options = { purpose: :cms_edit_mode, expires_at: }
+    signature = verifier.generate(cms_token.b, **signing_options).split("--").last
+
+    redirect_to access_code_url(
+      host: current_account.external_domain,
+      signature:,
+      expires_at: expires_at.to_i,
+      access_code: current_account.site_access_code,
+      return_to:,
+      cms: cms_mode)
+  end
+
+  def return_to
+    params.permit(:return_to)[:return_to]
+  end
+
+  def cms_mode
+    session[:cms]
+  end
```

Comment:
just a nitpick. should these be private?

Reply:
Sure: https://github.com/3scale/porta/pull/3938/commits/ef951103e639bf9303b4d0c54307308e17ac275a